sky/data/storage.py

```diff
@@ -843,47 +876,58 @@ def delete(self) -> None:
+
+        def get_file_sync_command(base_dir_path, file_names):
+            includes = ' '.join(
+                [f'--include "{file_name}"' for file_name in file_names])
+            sync_command = ('aws s3 sync --no-follow-symlinks --exclude="*" '
+                            f'{includes} {base_dir_path} '
+                            f's3://{self.name}')
+            return sync_command
+
+        def get_dir_sync_command(src_dir_path, dest_dir_name):
+            sync_command = ('aws s3 sync --no-follow-symlinks '
+                            f'{src_dir_path} '
+                            f's3://{self.name}/{dest_dir_name}')
+            return sync_command
+
+        # Generate message for upload
+        if len(source_path_list) > 1:
+            source_message = f'{len(source_path_list)} paths'
+        else:
+            source_message = source_path_list[0]
+
         with backend_utils.safe_console_status(
                 f'[bold cyan]Syncing '
-                f'[green]{self.source}[/] to [green]s3://{self.name}/[/]'):
-            # TODO(zhwu): Use log_lib.run_with_log() and redirect the output
-            # to a log file.
-            with subprocess.Popen(sync_command.split(' '),
-                                  stderr=subprocess.PIPE,
-                                  stdout=subprocess.DEVNULL) as process:
-                stderr = []
-                while True:
-                    line = process.stderr.readline()
-                    if not line:
-                        break
-                    str_line = line.decode('utf-8')
-                    stderr.append(str_line)
-                    if 'Access Denied' in str_line:
-                        process.kill()
-                        with ux_utils.print_exception_no_traceback():
-                            raise PermissionError(
-                                'Failed to upload files to '
-                                'the S3 bucket. The bucket does not have '
-                                'write permissions. It is possible that '
-                                'the bucket is public.')
-                returncode = process.wait()
-                if returncode != 0:
-                    stderr = '\n'.join(stderr)
-                    with ux_utils.print_exception_no_traceback():
-                        logger.error(stderr)
-                        raise exceptions.StorageUploadError(
-                            f'Upload to S3 failed for store {self.name} and '
-                            f'source {self.source}. Please check the logs.')
+                f'[green]{source_message}[/] to [green]s3://{self.name}/[/]'):
+            data_utils.parallel_upload(
+                source_path_list,
+                get_file_sync_command,
+                get_dir_sync_command,
+                self.name,
+                self.ACCESS_DENIED_MESSAGE,
+                create_dirs=create_dirs,
+                max_concurrent_uploads=_MAX_CONCURRENT_UPLOADS)
```

Comment:
Has a scalability test been tried for this? (i.e. what if we had 32 imagenet folders to upload to the bucket, does it kill the computer?) I checked online and it says that `aws s3 sync` by default uses 10 threads. 10*32 = 320 threads.

Reply:
It might be good to try a larger version in a smoke test, but I'm fine with leaving it out for now.

Reply:
Great point - I just tried with 20 file_mounts, seemed to work okay on my 2018 x1 yoga (i7-8650u, 16gb RAM).

Reply:
@romilbhardwaj After this PR, good idea to file a Git issue for scalability tests for Sky Storage!